README
                    DDS Benchmarking Environment


last updated 4/27/07


               Generating builds with MPC
               
Check out the MPC module from the SubVersion repository. It's
location is

https://svn.dre.vanaderbilt.edu/DOC/MPC/trunk

Move the contents of the 'trunk' directory up to the level of
the 'trunk' directory, and delete it. Set $MPC_ROOT in your
environment to the path of this new directory. Building the
entire DDS_Test directory include building DEEP and lots
of tests and examples - most of the latter depend on ACE
and some depend on TAO. If you have ACE+TAO installed,
then, from the directory this README file is in, type

$ACE_ROOT/mwc.pl -type <type> DDSTest.mwc

where <type> is 'make' or 'vc71' or 'vc8'. 

              SVN COMMANDS FOR THIS PROJECT:

Legend:
  #user = your dre.vanderbilt.edu username

To checkout a fresh copy:

svn co https://svn.dre.vanderbilt.edu/DOC/DDS_Test/trunk/ --username #user .

To update the files to the latest version from the repository:

svn update --username #user

To commit files to the repository:

svn commit --username #user -m "ChangeLogTag:`head -n1 ChangeLog`"



ChangeLog
Wed Sep 26 20:27:30 UTC 2007  Abdullah Sowayan  <abdullah.sowayan@lmco.com>

        * MPC/config/ndds_ts_defaults.mpb:
        * MPC/config/splice_ts_defaults.mpb:

          Removed extra colons in mpb files. The extra colons cause make warnings
          to appear while building. Thanks to Chad Elliott for pointing that out.

Wed Aug 22 18:12:07 UTC 2007  Abdullah Sowayan  <abdullah.sowayan@lmco.com>

        * MPC/config/nddslib.mpb:

          Enhanced mpb's file support for MSVC71

Fri May 25 16:46:35 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/doc/VendorSpecificIssues.html:

          Changes intended but overlooked in

          Thu Apr 26 16:59:11 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

Thu May 17 15:35:17 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/doc/VendorSpecificIssues.html:

          Added an item to the table about RTI DDS's proprietary
          ParticipantQos add-ons, some of which must be set in
          application code. Thanks to Hans van't Hag <hans dot
          vanthag at prismtech dot com> for reminding me.

Wed May 16 20:13:43 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/DEEP_RTI_DDS.mpc:
        * DEEP/DEEP_OpenDDS.mpc:
        * DEEP/DEEP_OpenSplice.mpc:

          Removed again the IDL processing steps. There is no clear
          way to accomplish the generation of a correct makefile
          that follows the directory structure and redirection of
          output in the existing build scripts, especially since
          the OpenSplice IDL processor has no option to redirect
          its output. Anyway, since IDL processing will be specific
          to each application, it will likely be removed from DEEP
          eventually.

Wed May 16 17:51:38 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * MPC/config/deepbase.mpb:

          Factored out 'exename' since it's the same for each MPC file.

        * DEEP/DEEP_RTI_DDS.mpc:
        * DEEP/DEEP_OpenDDS.mpc:
        * DEEP/DEEP_OpenSplice.mpc:

          Put back IDL processing steps for each vendor.

Wed May 16 16:12:51 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/DEEP.mwc:
        * DEEP/DEEP_RTI_DDS.mpc:
        * DEEP/DEEP_OpenSplice.mpc:

          Changes to these MPC-related files to get generated makefiles
          that are as close to the existing build scripts as possible.
          Thanks to Chad Elliott <elliott_c at ociweb dot com> for his
          help with MPC details.

Wed May  9 21:00:07 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/DEEP_RTI_DDS.mpc:

          Fixed typo in executable name.

Wed May  9 20:41:23 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/DEEP_RTI_DDS.mpc:
        * DEEP/DEEP_OpenSplice.mpc:
        * DEEP/DEEP_OpenDDS.mpc:

          Fixed some spelling case errors that showed up only on
          Linux. Also removed the OUTPUT_OPTION verbatim line, since
          it wasn't working as expected in generating a makefile,
          and added a CXX verbatim line for DEEP_OpenSplice.mpc,
          to get the version of g++ needed by OpenSplice, and an
          'install' line to put the executable into the directory
          needed by the DBE scripts. Thanks to Ming Xiong
          <ming dot xiong at vanderbilt dot edu> for his help with
          these makefile problems.

Mon May  7 01:35:55 UTC 2007  xiong,ming  <ming.xiong@vanderbilt.edu>

        * DEEP/readerWriter/deepTypes.idl

        Added MAX_MSG_LENGTH to octet sequence definition because
        the internal buffer allocated for serializing the sequence
        is based on this value, which means without which "bad
        things" might happen.

Sat Apr 28 14:38:44 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * DEEP/readerWriter/impl/TypedFunctions.cpp:
        * DEEP/readerWriter/impl/TypedFunctions.h:
        * DEEP/readerWriter/impl/ReaderWriterFactory_impl.cpp:
        * DEEP/scenario\impl/ActorSetting_impl.cpp:
        * DEEP/settingsParser/impl/SettingsParserDefault_impl.cpp:

          Removed module from IDL file regenerated code, and removed
          namespace 'deepTypes' from macro instantiations.

Sat Apr 28 01:42:31 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * MPC\config\MPC.cfg:
        * DEEP\DEEP.mpc:

          Removed these recently added files, reconfigured and
          added new files to get an easier procedure for doing
          MPC builds.

        * MPC\config\default.features:
        * MPC\config\deepbase.mpb:
        * DEEP\DEEP.mwc:
        * DEEP\DEEP_RTI_DDS.mpc:
        * DEEP\DEEP_OpenDDS.mpc:
        * DEEP\DEEP_OpenSplice.mpc:

          New files, separate MPC files for each flavor of DEEP
          now, and .mwc file dedicated just to DEEP, a
          default features file to enable the pickup of the
          Xerces base project, and a base .mpb project to
          factor out common stuff from all 3 builds.

        * README:
        * DEEP\README:

          Updated documentation to reflect the changes described
          above.

Fri Apr 27 21:46:45 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * MPC/config/MPC.cfg:

          New file, to be copied to $MPC_ROOT/config for use with
          MPC to build things in DDS_Test that require ACE and/or
          TAO.

        * README:

          Updated MPC information.

Fri Apr 27 21:15:25 UTC 2007  Jeff Parsons <j.parsons@vanderbilt.edu>

        * MPC\config\nddslib_with_idl.mpb:
        * MPC\config\spliceexe_with_idl.mpb:
        * MPC\config\splicelib.mpb:
        * MPC\config\splicelib_with_idl.mpb:
        * MPC\config\nddsexe_with_idl.mpb:
        * MPC\config\nddslib.mpb:

          New files, a better breakdown of functionality
          that provides every combination of lib vs
          executable and with/without typesupport processing.

        * DEEP\DEEP.mpc:


scripts/terminate.pl
#!/usr/bin/perl


use Env (DBE_SCRIPTS);
use Env (DBE_ROOT);
use Env (DBE_LOCAL_ROOT);
use Env (DBE_RESULTS_BACKUP);


require "$DBE_SCRIPTS/scripts.lib";

# AUTHOR'S NOTE!!!

# THIS SCRIPT WILL KILL ANY PERL TRANSLATION THAT IS CURRENTLY RUNNING
# ON THE NODE AND IS KILLABLE BY THE USER! THERE ARE NO OTHER EXCEPTIONS!

# if @ARGV == 0, then no test file was specified. test files provide us
# with information about the test that was run, so we can make intelligent
# killall calls. If nothing is specified, we blindly kill anything on 30-45.

&setEnvironment("$DBE_ROOT/settings/environment");
&setEnvironment("$DBE_ROOT/settings/ospl_environment");

$node;

if( @ARGV > 0 )
{
  if( @ARGV[0] eq "-h" )
  {
    print "\n" .
      "Usage:\n\n" .
      "  terminate.pl [-h] [-s signal] [-i impl] 
             [-n netSettings] [-q defaultQoS] [-type testType]
             [testSettings]\n";
    print "
  [-h] print usage information for this command.
  [-s signal] specifies the signal to kill processes with.
  [-i impl] specifies the implementation of the processes to kill.
  [-n netSettings] specifies network settings of test to kill.
  [-q defaultQoS] specifies QoS settings of test to kill.
  [-type testType] specifies the test type (throughput, etc.).
  [testSettings] specifies the main test settings file.\n\n";

    exit(0);
  }
}


if( @ARGV[0] eq '-s' )
{
  shift(@ARGV);
  $signal = '-s ' . shift(@ARGV);
}
else
{
  $signal = '-s 9';
}

if( @ARGV[0] eq '-n' )
{
  shift(@ARGV);
  $node = shift(@ARGV);
}

print "Killing all processes with signal: $signal\n";


if( @ARGV == 0 )
{
  for( $i = 30; $i <= 45; $i++ )
  {
    print "\nTerminating processes on blade$i.\n";
    system("ssh tczar\@blade$i.isislab.vanderbilt.edu 'sudo killall $signal -v perl'");
    system("ssh tczar\@blade$i.isislab.vanderbilt.edu 'sudo killall $signal -v DCPSInfoRepo'");
    system("ssh tczar\@blade$i.isislab.vanderbilt.edu 'sudo killall $signal -v subscriber'");
    system("ssh tczar\@blade$i.isislab.vanderbilt.edu 'sudo killall $signal -v publisher'");

    sleep(4);

    system("ssh tczar\@blade$i.isislab.vanderbilt.edu $DBE_SCRIPTS/ospl_stop");
  }
  exit(0);
}

if( $ARGV[0] eq "-h" || $ARGV[0] eq "-u" || $ARGV[0] eq "--help" )
{
  print "Script that terminates all processes started by benchmark.pl\n";
  print "Note that instances of benchmark.pl are NOT terminated.\n";
  print "  usage:  terminate.pl [testfile]\n";
}

# a test file must have been specified. let's see if it has anything for us.

%settings = ();

&readSettingsFromArgs();

$impl = lc($settings{'impl'});

if( $impl eq "ndds" )
{
}
elsif( $impl eq "tao" || $impl eq "tao-dds" || $impl eq "tao_dds"
     || $impl eq "taodds" )
{
  $impl = 'tao';
}
elsif( $impl eq "splice" || $impl eq "opensplice" || $impl eq "open-splice"
     || $impl eq "open_splice" )
{
  $impl = 'splice';
}



if( $impl ne "tao" )
{
  $settings{'repos'} = 0;
}

if( $settings{'nodelist'} eq "true" )
{

  print "Repos: " . $settings{'repos'} . " Pubs: " .  $settings{'publishers'}
      . " Subs: " . $settings{'subscribers'} . "\n";

  $i = $settings{'repos'} % $settings{'nodes'};
  $j = 0;
  for( $j = 0; $j < $settings{'publishers'}; $j++, 
                                            $i = ($i + 1)%$settings{'nodes'} )
  {
    $host = "";
    if( $settings{"node$i"} ne "" )
    {
      $host = $settings{"node$i"};
    }
    elsif( $settings{"pub$j"} ne "" )
    {
      $host = $settings{"pub$j"};
    }

    if( $host ne "" )
    {
      print "Killing all publishers on $host...\n";
      system("ssh tczar\@$host 'sudo killall $signal -v perl'");
      sleep(1);
      system("ssh tczar\@$host 'sudo killall $signal -v publisher'");

      if( $impl eq "tao" )
      {
        system("ssh tczar\@$host 'sudo killall $signal -v tao_pub'");
      }

      if( $impl eq "splice" )
      {
        sleep(2);
        print "ssh tczar\@$host 'sudo ospl stop'\n";
        system("ssh tczar\@$host $DBE_SCRIPTS/ospl_stop");
      }
    }
    else
    {
      die "One of your publishers (node$i or pub$j) is not setup properly.\n";
    }

  }

  $i = $settings{'publishers'} + $settings{'repos'};
  $i = $i % $settings{'nodes'};

  $j = 0;
  for( $j = 0; $j < $settings{'subscribers'}; $j++, 
                                              $i = ($i + 1)%$settings{'nodes'} )
  {
    $host = "";
    if( $settings{"node$i"} ne "" )
    {
      $host = $settings{"node$i"};
    }
    elsif( $settings{"sub$j"} ne "" )
    {
      $host = $settings{"sub$j"};
    }

    if( $host ne "" )
    {
      print "Killing all subscribers on $host...\n";
      system("ssh tczar\@$host 'sudo killall $signal -v perl'");
      sleep(1);
      system("ssh tczar\@$host 'sudo killall $signal -v subscriber'");

      if( $impl eq "tao" )
      {
        system("ssh tczar\@$host 'sudo killall $signal -v tao_sub'");
      }

      if( $impl eq "splice" )
      {
        sleep(2);
        system("ssh tczar\@$host $DBE_SCRIPTS/ospl_stop");
      }
    }
    else
    {
      die "One of your subscribers (node$i or sub$j) is not setup properly.\n";
    }
  }

  $i = 0;
  $j = 0;
  for( $j = 0; $j < $settings{'repos'}; $j++,
                                        $i = ($i+1)%$settings{'nodes'} )
  {
    $host = "";
    if( $settings{"node$i"} ne "" )
    {
      $host = $settings{"node$i"};
    }
    elsif( $settings{"repo$j"} ne "" )
    {
      $host = $settings{"repo$j"};
    }

    if( $host ne "" )
    {
      print "Killing all repos on $host...\n";
      system("ssh tczar\@$host 'sudo killall $signal -v perl'");
      sleep(1);
      system("ssh tczar\@$host 'sudo killall $signal -v DCPSInfoRepo'");

    }
    else
    {
      die "One of your repos (node$i or repo$j) is not setup properly.\n";
    }

  }
}
else
{
  die "If you provide a test file, it must have a node listing.\n" .
      " usage:   terminate.pl [testfile]\n";
}
